Source/WebCore/html/canvas/UsdModelLoader.swift

```diff
@@ -0,0 +1,678 @@
+            for l in nonNullLayouts {
+                var layout = WebDDVertexLayout(
+                    bufferIndex: Int32(l.bufferIndex),
+                    bufferOffset: Int32(l.bufferOffset),
+                    bufferStride: Int32(l.bufferStride)
+                )
+                results.append(layout)
+            }
+        }
+        return results
+    }
+
+    static func webUpdateMeshRequestFromUpdateMeshRequest(request: UpdateMeshRequest) -> WebUpdateMeshRequest {
+        var webRequest = WebUpdateMeshRequest(
+            partCount: request.partCount ?? 0,
+            parts: Converter.toWebSetParts(parts: request.parts),
+            renderFlags: Converter.toWebRenderFlags(flags: request.renderFlags),
+            vertices: Converter.toWebReplaceVertices(verts: request.vertices),
+            indices: request.indices,
+            transform: request.transform ?? simd_float4x4(),
+            materialIds: request.materialIds
+        )
+        Helper.setTransform(optionalChain: webRequest.instanceTransforms, arr: request.instanceTransforms, index: 0)
+
+        return webRequest
+    }
+
+    static func webAddMeshRequestFromAddMeshRequest(request: AddMeshRequest) -> WebAddMeshRequest {
+        WebAddMeshRequest(
+            indexCapacity: Int32(request.descriptor.indexCapacity),
+            indexType: Int32(request.descriptor.indexType.rawValue),
+            vertexBufferCount: Int32(request.descriptor.vertexBufferCount),
+            vertexCapacity: Int32(request.descriptor.vertexCapacity),
+            vertexAttributes: toWebVertexAttributes(attrs: request.descriptor.vertexAttributes),
+            vertexLayouts: toWebVertexLayouts(layouts: request.descriptor.vertexLayouts)
+        )
+    }
+}
+
+@objc
+@implementation
+extension WebUSDModelLoader {
+    @nonobjc
+    var renderer: USDModelLoader?
+    @nonobjc
+    var modelAdded: ((WebAddMeshRequest) -> (Void))?
+    @nonobjc
+    var modelUpdated: ((WebUpdateMeshRequest) -> (Void))?
+    @nonobjc
+    var modelAddedCallback: ((Int32) -> (Void))?
+
+    override init() {
+        super.init()
+
+        self.renderer = USDModelLoader(objcInstance: self)
+    }
+    @objc(setCallbacksWithModelAddedCallback:modelUpdatedCallback:)
+    func setCallbacksWithModelAddedCallback(
+        _ modelAddedCallback: @escaping ((WebAddMeshRequest) -> (Void)),
+        modelUpdatedCallback: @escaping ((WebUpdateMeshRequest) -> (Void))
+    ) {
+        self.modelAdded = modelAddedCallback
+        self.modelUpdated = modelUpdatedCallback
+    }
+
+    @objc
+    func loadModel(from url: URL) {
+        self.renderer?.loadModel(from: url)
+    }
+
+    fileprivate func updateMesh(webRequest: WebUpdateMeshRequest) {
+        Helper.logger.info("update! modelUpdated")
+        if let mu = modelUpdated {
+            mu(webRequest)
+        }
+    }
+
+    fileprivate func addMesh(
+        webRequest: WebAddMeshRequest,
+        callback: @escaping (Int32) -> Void = { result in
```

Comment:
this is a bit bizarre; i think just having two separate overloads would be better and much more readable

Reply:
This allows addMesh to be asynchronous.

I don't think two separate methods allows this.